### PR TITLE
HIVE-23454 : Querying hive table which has Materialized view fails with HiveAccessControlException

### PR DESCRIPTION
**Pull request by vineetgarg02**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 1]
```diff
@@ -1736,6 +1768,10 @@ public boolean validateMaterializedViewsFromRegistry(List<Table> cachedMateriali
       // Final result
       boolean result = true;
       for (Table cachedMaterializedViewTable : cachedMaterializedViewTables) {
+        if (!checkPrivillegeForMV(cachedMaterializedViewTable)) {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:1771]: `validateMaterializedViewsFromRegistry` is only called if the MV is coming from the registry. However, we need the authorization check in all cases, e.g., dummy registry.
You can simply call the new method from Calcite planner.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -1718,6 +1722,34 @@ public Table apply(org.apache.hadoop.hive.metastore.api.Table table) {
     }
   }
 
+  /**
+   * Validate if given materialized view has SELECT privileges for current user
+   * @param cachedMVTable
+   * @return false if user does not have privilege otherwise true
+   * @throws HiveException
+   */
+  private boolean checkPrivillegeForMV(final Table cachedMVTable) throws HiveException{
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:1731]: Can we move this to HiveMaterializedViewUtils?

There is also a typo: `Privillege`

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 3]
```diff
@@ -1718,6 +1722,34 @@ public Table apply(org.apache.hadoop.hive.metastore.api.Table table) {
...
+            .map(FieldSchema::getName)
+            .collect(Collectors.toList());
+
+    HivePrivilegeObject privObject = new HivePrivilegeObject(cachedMVTable.getDbName(),
+        cachedMVTable.getTableName(), colNames);
+    List<HivePrivilegeObject> privObjects = new ArrayList<HivePrivilegeObject>();
+    privObjects.add(privObject);
+    try {
+      SessionState.get().getAuthorizerV2().
+          checkPrivileges(HiveOperationType.QUERY, privObjects, privObjects, new HiveAuthzContext.Builder().build());
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:1743]: Can we check the privileges for all MVs used by the query at once so we do not need multiple round trips?

**Comment by vineetgarg02** [comment]
@jcamachor  I have addressed your review comments in latest update.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 4]
```diff
@@ -2280,6 +2280,15 @@ private RelNode applyMaterializedViewRewriting(RelOptPlanner planner, RelNode ba
         return calcitePreMVRewritingPlan;
       }
 
+      try {
+        if (!HiveMaterializedViewUtils.checkPrivilegeForMV(materializedViewsUsedAfterRewrite)) {
+          // if materialized views do not have appropriate privilges, we shouldn't be using them
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:2285]: nit. typo

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveMaterializedViewUtils.java [hunk 5]
```diff
@@ -347,6 +355,38 @@ public static RelNode copyNodeNewCluster(RelOptCluster optCluster, RelNode node)
     }
   }
 
+  /**
+   * Validate if given materialized view has SELECT privileges for current user
+   * @param cachedMVTable
+   * @return false if user does not have privilege otherwise true
+   * @throws HiveException
+   */
+  public static boolean checkPrivilegeForMV(List<Table> cachedMVTableList) throws HiveException {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveMaterializedViewUtils.java:364]: Can we use full name: `checkPrivilegeForMaterializedViews` ?

**Review by jcamachor — APPROVED** [review]
Left a couple of minor comments. Other than that, LGTM. +1

**Comment by jcamachor** [comment]
We should probably create a follow-up JIRA to check authorization before triggering the rewriting algorithm. If the compilation overhead to check every MV that is applicable to the given query is unacceptable, permissions could possibly be kept in the HS2 registry and refreshed periodically in the background, then verified after rewriting, which would at least decrease the number of authorization failures. @vineetgarg02 , can you create a JIRA for this?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveMaterializedViewUtils.java (hunk 5 above)
Comment (vineetgarg02) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveMaterializedViewUtils.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 4 above)
Comment (vineetgarg02) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java]: Fixed

**Comment by vineetgarg02** [comment]
> We should probably create a follow-up JIRA to check authorization before triggering the rewriting algorithm. If the compilation overhead to check every MV that is applicable to the given query is unacceptable, permissions could possibly be kept in the HS2 registry and refreshed periodically in the background, then verified after rewriting, which would at least decrease the number of authorization failures. @vineetgarg02 , can you create a JIRA for this?

@jcamachor  I have created a jira at https://issues.apache.org/jira/browse/HIVE-24140

**Pull request merged**